### ISSUE TITLE
fix actix-web version

### ISSUE THREAD

**Issue by joestubbs** [joestubbs/tagent#3]
It seems major versions of actix web are only compatible with specific tokio runtime versions; specifically, it looks like actix 4 (in beta) is compatible with tokio 1.2, but actix 3 is only compatible with tokio 0.2. See, for instance, this issue [1]. I am getting a panic when trying to use my JWT parser (it actually cannot make the http request to Tapis to fetch the public key). The error is "must be called from the context of a Tokio 1.x runtime...", similar to the issue above. The tagent setup at the moment is actix-web 0.3 but it appears to have pulled in tokio 1.2, probably from some other dependency. Interestingly, when I specify tokio = { version = "0.2", features = ["time"] }in the cargo.toml file, I still get the same error about a 1.2 tokio runtime.

On the other hand, when I switch to actix 4, I get compiler errors, probably because the API has changed. Probably switching to actix 4 is the right move, but I think the main docs site [2] is still for actix 3. For instance, I implemented the Either type for returning different return types from handlers as documented at [3], and my implementation works with actix 3 but does not compile with actix.

We can discuss actix-web 3 vs 4 in this issue and decide which to use.

[1] https://github.com/actix/actix-web/issues/2155
[2] https://actix.rs/docs
[3] https://actix.rs/docs/handlers/

**Comment by joestubbs** [comment]
I haven't been able to find docs for 4.0 but the examples in this repository seem to be updated to use 4: https://github.com/actix/examples
Working off these examples to upgrade to 4 seems like a doable approach.

**Comment by waltermoreira** [comment]
Here's a list of the changes in v4: https://github.com/actix/actix-web/releases?q=v4.0.0&expanded=true, but it's long, might be unwieldly to go over it.

The examples look like a great start, since they are updated to the new version.

My vote would be to take the plunge to v4. A potential idea would be to start posting as issues all the errors we get while compiling the code as-is in v4, and then we can split the work of fixing them. Do you think it'll be a lot of work? I honestly haven't looked in detail yet, but I'd love to help in the porting, if that's the direction you want to go.

**Comment by joestubbs** [comment]
Awesome! I was thinking that taking the plunge to v4 might be the right move as well. Let's do it! I actually started looking at the compiler errors from moving to 0.4 today. 

Yes, let's try for writing down the tasks coming from the errors. I'd be interested in your input here. One challenge is the how the changes kind of cascade one into the other. 

For example, one problem I think is the use of the Either return type in the handlers, to return different kinds of responses, which doesn't compile in 4. Another problem is the use of the NamedFile type in the download endpoint response. I tried to start fixing these, and it led me to the conclusion that maybe we need custom error types. I had already been thinking this, kind of in the abstract, because it seems from my (limited) reading that this is the Rust way, but for actix 4 I am wondering if will also just be the best way to implement these error responses (and fix that Either stuff). I am kind of basing that on this example here: https://github.com/actix/examples/blob/master/basics/error_handling/src/main.rs

Let me know what you think.. it might be the kind of work best done in a work session together, but I am also happy for either of us to make issues with things as we hit them.

**Comment by joestubbs** [comment]
I was getting some very odd errors, but I figured out that some of the versions specified in Cargo.toml for some of the actix web "extras" crates were causing v3 of actix-web itself to get pulled in, even though I specified `actix-web = "4.0.0-beta.5"` in the Cargo.toml file. For example, I could see something like this in the Cargo.lock file:
`[[package]]

name = "actix-multipart"
version = "0.3.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "774bfeb11b54bf9c857a005b8ab893293da4eaff79261a66a9200dab7f5ab6e3"
dependencies = [
 "actix-service 1.0.6",
 "actix-utils 2.0.0",
 **"actix-web 3.3.3",**
 "bytes 0.5.6",
 "derive_more",
 "futures-util",
 "httparse",
 "log",
 "mime",
 "twoway",
]`

In other words, both actix web v4 and v3 were getting pulled in! So, first order of business is to update all of the extras packages to the versions corresponding to actix web v4. I am going to try to at least get that done this evening.

**Comment by joestubbs** [comment]
Fixing the multiple actix web versions made this much easier (imagine that), and last night I was able to get the code to compile in v4. It still doesn't run correctly, as there is an issue with the app state -- it appears v4 made some changes to the API that break things in ways that don't show up at compile time. I don't expect that to be too hard though, so we might be pretty close to having a working version in v4 (though I could hit additional issues after fixing the app state). 
Even still, there are a number of things that will need to be improved -- for instance, custom errors. Plenty of good items for discussion.

**Comment by waltermoreira** [comment]
Great to hear! 🎉

**Comment by waltermoreira** [comment]
I'm curious and eager to start looking at these changes for v4 😄 . Are you working on a branch that hasn't been pushed?

**Comment by joestubbs** [comment]
Ah, you're right! I was just telling my student it is important to push to the remote regularly, even before code is ready, if for nothing else so you have a backup, but I was not doing it myself. :)

I just pushed branch v4, so check that out. Still haven't merged those latest commit from main yet but the code does compile.
